### PR TITLE
gnrc/netif: add gnrc_netif_ipv6_wait_for_global_address()

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This adds a function to wait until a global address becomes available.

This is useful if a node wants to perform some action involving the network after wake-up / start-up.
It blocks until an uplink connection is available or the user specified timeout expired.

### Testing procedure

A new test was added as  `tests/gnrc_netif_ipv6_wait_for_global_address`

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/pull/15415#issuecomment-724653275

### CODE REVIEW

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 1]
```diff
@@ -52,6 +52,10 @@
 #define ENABLE_DEBUG 0
 #include "debug.h"
 
+#ifndef CONFIG_NETIF_NUM_MAX
+#define CONFIG_NETIF_NUM_MAX 2
+#endif
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:57]: Unrelated?

Comment (benpicco) [reply sys/net/gnrc/netif/gnrc_netif.c:57]: Unfortunately not.
I need a `msg_bus_entry_t` for each interface.

I could also use VLA instead :fearful:

Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c:57]: > Unfortunately not.
> I need a `msg_bus_entry_t` for each interface.

But we did so much work to not have this bound by a maximum size and it led to problems in the past already :-/. @jia200x what do you think? All in all, I would prefer it to be default 1 then, since that is what most our nodes usually have.

> I could also use VLA instead fearful

https://en.wikipedia.org/wiki/Vla?

Comment (benpicco) [reply sys/net/gnrc/netif/gnrc_netif.c]: > > I could also use VLA instead fearful
> 
> https://en.wikipedia.org/wiki/Vla?

yummy, but I meant 

https://en.wikipedia.org/wiki/Variable-length_array#C99

They should generally be avoided because of their unpredictable nature and no way to check if stack space was exhausted.
However, this might actually be a valid use case: `msg_bus_entry_t` is only 12 byte and the number of netifs is also small, so it's unlikely that this will overflow the stack.

**Comment by benpicco** [comment]
`gnrc_netif_ipv6_wait_for_prefix()` is catchy, but a bit misleading since we are actually waiting for an address.
`gnrc_netif_ipv6_wait_for_global_address()` would be an option. Or `gnrc_netif_ipv6_wait_for_uplink()`?

**Comment by miri64** [comment]
> `gnrc_netif_ipv6_wait_for_prefix()` is catchy, but a bit misleading since we are actually waiting for an address.
> `gnrc_netif_ipv6_wait_for_global_address()` would be an option. Or `gnrc_netif_ipv6_wait_for_uplink()`?

`gnrc_netif_wait_for_glb_addr()` if brevity is desired? `wait_for_uplink()` is as (or even more) misleading as `wait_for_prefix()`.

File: sys/net/gnrc/netif/gnrc_netif.c (hunk 1 above)
Comment (jia200x) [reply sys/net/gnrc/netif/gnrc_netif.c]: I'm fine with it, since `gnrc_netif_numof` will always resolve to a relatively small number.

> However, this might actually be a valid use case: msg_bus_entry_t is only 12 byte and the number of netifs is also small, so it's unlikely that this will overflow the stack.

In fact, a big-enough compile time value might also overflow the stack. So let's go with it

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 2]
```diff
@@ -1298,6 +1298,105 @@ int gnrc_netif_ipv6_add_prefix(gnrc_netif_t *netif,
 out:
     return res;
 }
+
+#if IS_USED(MODULE_GNRC_NETIF_BUS)
+static bool _has_global_addr(gnrc_netif_t *netif)
+{
+    ipv6_addr_t ipv6_addrs[CONFIG_GNRC_NETIF_IPV6_ADDRS_NUMOF];
+    int res = gnrc_netapi_get(netif->pid, NETOPT_IPV6_ADDR, 0, ipv6_addrs,
+                              sizeof(ipv6_addrs));
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:1307]: Since this is within `gnrc_netif.c` I think it is fine to use `gnrc_netif_acquire()`/`gnrc_netif_release()` in this function and access the IPv6 address array of the netif directly. Saves some cycles and memory.

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 3]
```diff
@@ -41,9 +41,7 @@
 #include "fmt.h"
 #include "log.h"
 #include "sched.h"
-#if (CONFIG_GNRC_NETIF_MIN_WAIT_AFTER_SEND_US > 0U)
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:44]: I think there was a problem with an all-out include, that's why I `#if`'d it. Maybe add `|| IS_USED(MODULE_GNRC_NETIF_BUS)`?

Comment (benpicco) [reply sys/net/gnrc/netif/gnrc_netif.c:44]: Not sure why a header file would cause trouble - Murdock seems fine with it (except for insufficient memory boards for the new test).

I would try to avoid `#ifdefs` where they are not needed to not clutter the code

Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c:44]: Ok, if murdock is happy NVM.

Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c:44]: > Ok, if murdock is happy NVM.

Ad https://github.com/RIOT-OS/RIOT/commit/9db92f0f81921229c28b2e57b2af919a5212f6d4, apparently it was not?

Comment (benpicco) [reply sys/net/gnrc/netif/gnrc_netif.c:44]: Yes turns out the include causes problems if the board does not support xtimer (IMHO that should be fixed in the header file)

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 4]
```diff
@@ -1298,6 +1298,107 @@ int gnrc_netif_ipv6_add_prefix(gnrc_netif_t *netif,
...
+    bool has_global = false;
+
+    if (netif) {
+        if (_has_global_addr(netif)) {
+            return true;
+        }
+
+        msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);
+        msg_bus_attach(bus, &subs[count]);
+        msg_bus_subscribe(&subs[count], GNRC_IPV6_EVENT_ADDR_VALID);
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:1337]: This could be a `static` function, ...

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 5]
```diff
@@ -1298,6 +1298,107 @@ int gnrc_netif_ipv6_add_prefix(gnrc_netif_t *netif,
...
+        /* subscribe to all interfaces */
+        while ((netif = gnrc_netif_iter(netif))) {
+            msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);
+
+            if (_has_global_addr(netif)) {
+                has_global = true;
+            }
+
+            msg_bus_attach(bus, &subs[count]);
+            msg_bus_subscribe(&subs[count], GNRC_IPV6_EVENT_ADDR_VALID);
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:1350]: ... That would also be called here.

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 6]
```diff
@@ -1298,6 +1298,107 @@ int gnrc_netif_ipv6_add_prefix(gnrc_netif_t *netif,
...
+        /* no interfaces */
+        if (count == 0) {
+            return false;
+        }
+    }
+
+    /* don't wait if we have an address already */
+    if (has_global) {
+        goto out;
+    }
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:1364]: `!has_global` could be the `while`-condition of the loop below and spare us a `goto` ;-).

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 7]
```diff
@@ -1298,6 +1298,107 @@ int gnrc_netif_ipv6_add_prefix(gnrc_netif_t *netif,
...
+            DEBUG_PUTS("gnrc_netif: timeout waiting for prefix");
+            break;
+        }
+
+        if (ipv6_addr_is_link_local(m.content.ptr)) {
+            DEBUG_PUTS("gnrc_netif: got link-local address");
+        } else {
+            DEBUG_PUTS("gnrc_netif: got global address");
+            has_global = true;
+            break;
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:1379]: And this break would become unnecessary as well :-).

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 8]
```diff
@@ -1298,6 +1298,107 @@ int gnrc_netif_ipv6_add_prefix(gnrc_netif_t *netif,
...
+        }
+    }
+
+out:
+    /* called with a given interface */
+    if (count == 0) {
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:1385]: Then why not use `netif != NULL` as a check here. For the `while` loop that subscribes to all interfaces above, `netif == NULL` is the pre- and post-condition, so `netif != NULL` would indicate that `netif` was set by the caller.

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 9]
```diff
@@ -1298,6 +1298,107 @@ int gnrc_netif_ipv6_add_prefix(gnrc_netif_t *netif,
...
+        msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);
+        msg_bus_detach(bus, &subs[count]);
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:1387]: Also could go into a function...

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 10]
```diff
@@ -1298,6 +1298,107 @@ int gnrc_netif_ipv6_add_prefix(gnrc_netif_t *netif,
...
+        return has_global;
+    }
+
+    /* unsubscribe all */
+    count = 0;
+    netif = NULL;
+    while ((netif = gnrc_netif_iter(netif))) {
+        msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);
+        msg_bus_detach(bus, &subs[count++]);
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:1396]: Same here

Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c:1396]: \+ I see some potential for breakage here:

`netif` and the `subs` element indirectly relate via the `count` variable. If some later change messes with `count`, this could go wrong. Is there maybe some more direct relation between `netif` and the `subs` element that can be made?

File: tests/gnrc_netif_ipv6_wait_for_global_address/main.c [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    _add_delayed_addr(&netif_test[1]);
+    TEST_ASSERT(!gnrc_netif_ipv6_wait_for_global_address(&netif_test[0], 20));
+
+    /* event when adding addr to any interface */
+    _add_delayed_addr(&netif_test[0]);
+    TEST_ASSERT(gnrc_netif_ipv6_wait_for_global_address(NULL, 20));
+
+    /* event when adding addr to specified interface */
+    _add_delayed_addr(&netif_test[0]);
+    TEST_ASSERT(gnrc_netif_ipv6_wait_for_global_address(&netif_test[0], 20));
```
Comment (miri64) [inline tests/gnrc_netif_ipv6_wait_for_global_address/main.c:91]: Maybe also add a check that `gnrc_netif_ipv6_wait_for_global_address()` blocked for the expected amount of time. While this also somewhat would test the underlying timers (which, granted, might not be desirable), it also would test the blocking behavior of `gnrc_netif_ipv6_wait_for_global_address()`.

**Review by miri64 — CHANGES_REQUESTED** [review]

File: tests/gnrc_netif_ipv6_wait_for_global_address/main.c (hunk 11 above)
Comment (benpicco) [reply tests/gnrc_netif_ipv6_wait_for_global_address/main.c]: added

File: sys/net/gnrc/netif/gnrc_netif.c (hunk 10 above)
Comment (benpicco) [reply sys/net/gnrc/netif/gnrc_netif.c]: something like 

```C
struct netif_sub {
    gnrc_netif_t *netif;
    msg_bus_entry_t sub;
}
```
?

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 12]
```diff
@@ -1320,35 +1320,41 @@ static bool _has_global_addr(gnrc_netif_t *netif)
     return has_global;
 }
 
+static void _netif_bus_attach_and_subscribe(gnrc_netif_t *netif, msg_bus_entry_t *sub)
+{
+    msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);
+    msg_bus_attach(bus, sub);
+    msg_bus_subscribe(sub, GNRC_IPV6_EVENT_ADDR_VALID);
+}
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:1328]:
```suggestion
static void _netif_bus_attach_and_subscribe_addr_valid(gnrc_netif_t *netif, msg_bus_entry_t *sub)
{
    msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);
    msg_bus_attach(bus, sub);
    msg_bus_subscribe(sub, GNRC_IPV6_EVENT_ADDR_VALID);
}
```

to reflect that you subscribe to the `GNRC_IPV6_EVENT_ADDR_VALID`?

Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c:1328]: Alternatively, you could also provide the event-type as a parameter, to make this function more re-usable.

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 13]
```diff
@@ -1376,24 +1377,19 @@ bool gnrc_netif_ipv6_wait_for_global_address(gnrc_netif_t *netif,
...
 
-out:
     /* called with a given interface */
-    if (count == 0) {
-        msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);
-        msg_bus_detach(bus, &subs[count]);
+    if (netif != NULL) {
+        _netif_bus_detach(netif, &subs[0]);
         return has_global;
     }
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c]:
```suggestion
    }
    else {
```

IMHO the less early returns, the better (and in case the optimizer gets confused it might also lead to smaller code size). Also I think that makes the code easier to understand: The if branch handles the "`netif` is set" case the else branch the "`netif` is unset".

**Review by miri64 — CHANGES_REQUESTED** [review]

File: sys/net/gnrc/netif/gnrc_netif.c (hunk 12 above)
Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: Can you please provide your reasoning for going for my first suggestion? After thinking about it, I think I like my alternative more.

File: sys/net/gnrc/netif/gnrc_netif.c (hunk 10 above)
Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: That would require iterating over the array every look-up, right? I think in that case I would prefer the current approach.

File: sys/net/gnrc/netif/gnrc_netif.c (hunk 12 above)
Comment (benpicco) [reply sys/net/gnrc/netif/gnrc_netif.c]: less code on the caller's side and there is only a single event type implemented currently.

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 14]
```diff
@@ -1298,6 +1298,103 @@ int gnrc_netif_ipv6_add_prefix(gnrc_netif_t *netif,
...
+        unsigned count = 0;
+
+        /* subscribe to all interfaces */
+        while ((netif = gnrc_netif_iter(netif))) {
+            if (_has_global_addr(netif)) {
+                has_global = true;
+            }
+
+            _netif_bus_attach_and_subscribe_addr_valid(netif, &subs[count]);
+            ++count;
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:1359]: wrt https://github.com/RIOT-OS/RIOT/pull/16824#discussion_r710076990, maybe tying `count` and `netif` somewhat closer together using C-syntax will prevent future breakage:

```suggestion
        /* subscribe to all interfaces */
        for (unsigned count = 0;
             (netif = gnrc_netif_iter(netif));
             count++) {
            if (_has_global_addr(netif)) {
                has_global = true;
            }

            _netif_bus_attach_and_subscribe_addr_valid(netif, &subs[count]);
```

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 15]
```diff
@@ -1298,6 +1298,103 @@ int gnrc_netif_ipv6_add_prefix(gnrc_netif_t *netif,
...
+
+    /* called with a given interface */
+    if (netif != NULL) {
+        _netif_bus_detach(netif, &subs[0]);
+    } else {
+        /* unsubscribe all */
+        unsigned count = 0;
+        while ((netif = gnrc_netif_iter(netif))) {
+            _netif_bus_detach(netif, &subs[count++]);
+        }
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c:1392]: Same here

```suggestion
        for (unsigned count = 0;
             (netif = gnrc_netif_iter(netif));
             count++) {
            _netif_bus_detach(netif, &subs[count]);
        }
```

**Review by miri64 — CHANGES_REQUESTED** [review]

File: sys/net/gnrc/netif/gnrc_netif.c (hunk 12 above)
Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: Not sure, I would call one more parameter "more code", but ok.

File: tests/gnrc_netif_ipv6_wait_for_global_address/main.c [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+{
+    unsigned now;
+
+    TEST_ASSERT_EQUAL_INT(TEST_NETIF_NUMOF, gnrc_netif_numof());
+
+    /* no event when adding addr to other interface */
+    _add_delayed_addr(&netif_test[1]);
+    now = xtimer_now_usec();
+    TEST_ASSERT(!gnrc_netif_ipv6_wait_for_global_address(&netif_test[0], 20));
+    TEST_ASSERT_EQUAL_INT(20, (xtimer_now_usec() - now + 500) / 1000);
```
Comment (miri64) [inline tests/gnrc_netif_ipv6_wait_for_global_address/main.c:87]:
```suggestion
    TEST_ASSERT(20 <= ((xtimer_now_usec() - now) / 1000));
```

Using `EQUAL_INT` is a bit optimistic IMHO, so rather check if `20 <=` the time difference (I used yoda style in the suggestion, to make the change more clearer, feel free to use `<= 20` instead ;-)).

File: tests/gnrc_netif_ipv6_wait_for_global_address/main.c [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+    /* event when adding addr to any interface */
+    _add_delayed_addr(&netif_test[0]);
+    TEST_ASSERT(gnrc_netif_ipv6_wait_for_global_address(NULL, 20));
```
Comment (miri64) [inline tests/gnrc_netif_ipv6_wait_for_global_address/main.c:91]: This one doesn't have a delay?

File: tests/gnrc_netif_ipv6_wait_for_global_address/main.c [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+    /* event when adding addr to specified interface */
+    _add_delayed_addr(&netif_test[0]);
+    TEST_ASSERT(gnrc_netif_ipv6_wait_for_global_address(&netif_test[0], 20));
```
Comment (miri64) [inline tests/gnrc_netif_ipv6_wait_for_global_address/main.c:95]: If I understand this correctly, there should be a delay here as well ;-).

**Review by miri64 — CHANGES_REQUESTED** [review]
Only some test hardening left.

File: tests/gnrc_netif_ipv6_wait_for_global_address/main.c [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+{
+    const ipv6_addr_t addr =  {{ 0x20, 0x01, 0x0d, 0xd8,
+                                 0x00, 0x00, 0x00, 0x00,
+                                 0x00, 0x00, 0x00, 0x00,
+                                 0x00, 0x00, 0x00, 0x01 }};
+
+    xtimer_msleep(10);
+    gnrc_netif_ipv6_addr_add(netif, &addr, 64, GNRC_NETIF_IPV6_ADDRS_FLAGS_STATE_VALID);
+    xtimer_msleep(5);
+    gnrc_netif_ipv6_addr_remove(netif, &addr);
```
Comment (miri64) [inline tests/gnrc_netif_ipv6_wait_for_global_address/main.c:51]: It would be cleaner, if this removial would be done in a `tear_down` function, and each test case being, well, their own test case (i.e. in a separate function).

File: tests/gnrc_netif_ipv6_wait_for_global_address/main.c [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+    _add_delayed_addr(&netif_test[1]);
+    now = xtimer_now_usec();
+    TEST_ASSERT(!gnrc_netif_ipv6_wait_for_global_address(&netif_test[0], 20));
+    TEST_ASSERT(20 <= ((xtimer_now_usec() - now) / 1000));
+
+    /* event when adding addr to any interface */
+    now = xtimer_now_usec();
+    _add_delayed_addr(&netif_test[0]);
+    TEST_ASSERT(gnrc_netif_ipv6_wait_for_global_address(NULL, 20));
+    TEST_ASSERT(((xtimer_now_usec() - now) / 1000) < 20);
```
Comment (miri64) [inline tests/gnrc_netif_ipv6_wait_for_global_address/main.c:93]:
```suggestion
    TEST_ASSERT(((xtimer_now_usec() - now) / 1000) <= 10);
```

? At least that's the time frame the `_add_delayed_addr()` adds the address in, right?

File: tests/gnrc_netif_ipv6_wait_for_global_address/main.c [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+{
+    unsigned now;
+
+    TEST_ASSERT_EQUAL_INT(TEST_NETIF_NUMOF, gnrc_netif_numof());
+
+    /* no event when adding addr to other interface */
+    _add_delayed_addr(&netif_test[1]);
+    now = xtimer_now_usec();
+    TEST_ASSERT(!gnrc_netif_ipv6_wait_for_global_address(&netif_test[0], 20));
+    TEST_ASSERT(20 <= ((xtimer_now_usec() - now) / 1000));
```
Comment (miri64) [inline tests/gnrc_netif_ipv6_wait_for_global_address/main.c:87]: Please use either yoda style or "normal" style for the assert. Don't mix them...

```suggestion
    TEST_ASSERT(((xtimer_now_usec() - now) / 1000) >= 20);
```

File: tests/gnrc_netif_ipv6_wait_for_global_address/main.c [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+    now = xtimer_now_usec();
+    _add_delayed_addr(&netif_test[0]);
+    TEST_ASSERT(gnrc_netif_ipv6_wait_for_global_address(NULL, 20));
+    TEST_ASSERT(((xtimer_now_usec() - now) / 1000) < 20);
+
+    /* event when adding addr to specified interface */
+    now = xtimer_now_usec();
+    _add_delayed_addr(&netif_test[0]);
+    TEST_ASSERT(gnrc_netif_ipv6_wait_for_global_address(&netif_test[0], 20));
+    TEST_ASSERT(((xtimer_now_usec() - now) / 1000) < 20);
```
Comment (miri64) [inline tests/gnrc_netif_ipv6_wait_for_global_address/main.c:99]: Same here, as for the other non-timeout test.

```suggestion
    TEST_ASSERT(((xtimer_now_usec() - now) / 1000) <= 10);
```

**Review by miri64 — CHANGES_REQUESTED** [review]

File: tests/gnrc_netif_ipv6_wait_for_global_address/main.c (hunk 20 above)
Comment (benpicco) [reply tests/gnrc_netif_ipv6_wait_for_global_address/main.c:93]: `_add_delayed_addr()` waits 10ms for adding the prefix, so it should be

    10 <= diff < 20

IMHO checking that we returned before the timeout expired should be enough though.

File: tests/gnrc_netif_ipv6_wait_for_global_address/main.c [hunk 23]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+static void test_wait_success(void)
+{
+    unsigned now;
+
+    /* event when adding addr to specified interface */
+    now = xtimer_now_usec();
+    _add_delayed_addr(&netif_test[0]);
+    TEST_ASSERT(gnrc_netif_ipv6_wait_for_global_address(&netif_test[0], 20));
```
Comment (miri64) [inline tests/gnrc_netif_ipv6_wait_for_global_address/main.c:100]: Last thing (hopefully): This can be its own function now and called at the end of the `success` test cases:

```C
static void _assert_wait_blocks(gnrc_netif_t *add_netif,
                                gnrc_netif_t *wait_netif,
                                uint32_t timeout)
{
    uint32_t now = xtimer_now_usec();

    _add_delayed_addr(add_netif);
    TEST_ASSERT(gnrc_netif_ipv6_wait_for_global_address(wait_netif,
                                                        timeout));
    TEST_ASSERT(((xtimer_now_usec() - now) / US_PER_MS) < timeout);
}
```

(Sorry I did not include the last line into my commenting... Github did not let me comment on that line range for some reason...)

Comment (benpicco) [reply tests/gnrc_netif_ipv6_wait_for_global_address/main.c]: Ok

**Review by miri64 — APPROVED** [review]
ACK! Please squash

**Comment by miri64** [comment]
Toggled murdock, since we need https://github.com/RIOT-OS/RIOT/pull/16867 in for this PR to be mergable.
